bitcoin/src/util/from_hex_str.rs

```diff
@@ -0,0 +1,44 @@
+
+//! FromHexStr
+//! Functionality to allow types to take hex strings as input
+
+#[cfg(feature = "no-std")]
+use crate::prelude::String;
+
+#[cfg(feature = "no-std")]
+use crate::alloc::string::ToString;
+
+
+/// Hex Parsing error
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[non_exhaustive]
+pub enum FromHexError {
+    /// Parsing Error
+    ParseHex(String),
```

Comment:
I noticed you're converting inner error to a string. Please don't do that - that loses a bunch of information. I suggest one of these options:

```rust
pub enum FromHexError<E> {
    /// The input was not a valid hex string.
    ParseHex(E),

    /// The input is missing `0x` prefix.
    ///
    /// The field `0` contains the input that was attempted to be parsed.
    MissingPrefix(String),
}
```

```rust
pub enum FromHexError<E> {
    /// The input is not a valid hex string.
    ///
    /// The field `0` contains the input that was attempted to be parsed.
    // The field `1` contains the inner error.
    ParseHex(String, E),
    /// The input is missing `0x` prefix.
    ///
    /// The field `0` contains the input that was attempted to be parsed.
    MissingPrefix(String),
}
```

The former requires a newtype for each error but may be better when users use `no_prefix` method because then they don't have to deal with the  `MissingPrefix` branch. IOW the former is nicer to consumers but will be a bit annoying for us unless we come up with some more ideas to improve it.

Reply:
I've figured we don't need the `MissingPrefix` error variant when we can just check if it starts with the prefix and pass `s` as is to the no prefix method. The only way for the no prefix method to fail is the call to from_str_radix which raises the question if we need the custom error type at all (which I still think we do)

Reply:
basically, any valid hex representation (with prefix or not) will work and only needs the implementation for conversion of the non-prefix path. Which I think makes it easier